
plotfun.TRs.only   <- function(samples=NA, visfrom = NA, visto = NA, vline= NULL,  vline2 = NULL, # breakseq = 1500,
                               breakseq=NULL, ylim=NULL, cagefr.clust=NULL, 
                               annot_fill_column = annot_fill_column, palette=colorvec, tr.palette = colorvec,
                               TR.merged.data=TR.merged.data.sub, plot.data=plot.sum.sub, 
                               transcript.sizes = data.frame(
                                 transcript.label.size = 1.25,
                                 exon.rect_height = 0.1, exon.rect.linewidth=0.01,
                                 intron.linetype = 'dashed', intron.linewidth = 1),
                               ...) {
  
  FigY    <- plot.genome.region(visfrom=visfrom,	visto=visto, samples = samples,
                                facet_cropF = facet_nested(rows=vars(hpi), scales=scales, drop=T),
                                plot.data=plot.data, #[plot.sum$strand == '+',], 
                                genome.and.transcripts = T, transcripts.only = T,
                                annot_fill_column = annot_fill_column,
                                add.genome.plot=T, genome.only = F,
                                add.transcripts.plot =T, TR.merged.data=TR.merged.data, add.CAGE_significance.to.TRs=F, 
                                transcript.sizes = transcript.sizes, tr.size.multip = 0.75, transcript.plot.scale=1.25, TR.y.add = 0.5,
                                facet_genes = NULL, facet_TR = NULL,
                                gene.plotdata=gene.plotdata, genome=genome, prime = prime,
                                sum.counts.in.window=T, bin_width = bin_width, add.all.pos=T, y.thresh=NA,
                                gene.label=T, gene_name_col = 'gene_name', tr.palette = tr.palette,
                                add.unstranded=T, 
                                add.feature=T, 
                                force.all.gene.up.and.down = T,
                                force.all.gene.down=F, angle = angle,
                                add.cageTSS=F, cagefr.clust=cagefr.clust, cagefr.clust.dist=3,
                                cluster.aes = aes(xmin = region_start, xmax = region_end, y = y, forward = orientation, fill = strand), 
                                add.cageTSS.features = F,
                                crop.FALSE = crop.FALSE, scales = scales,
                                geom    = geom_area(aes(x=prime5, y = count, fill=strand, color = strand), 
                                                    position = 'identity', alpha=alpha$cov_geom, color = "black", linewidth=0.5),
                                #geom    = geom_point(aes(x=prime5, y = count, color=strand), alpha=alpha$cov_geom, size=0.1),
                                palette=colorvec, alpha=alpha, genomplot.scale=genomplot.scale,
                                gene.sizes = gene.sizes, sizes = sizes,
                                y.multip   = 3, y.log10 = F, ylim=ylim, #ybreaks = -5:5, 
                                ylims.gene = c(-2, 6),
                                force.gene.y = F, 
                                gene.label.col='black',
                                legend.position.prime='none', plot.title = NA, #letters[2], #paste0(virus, '_', Ori),
                                breakseq=breakseq, 
                                margins=unit(c(0.5, 0.5, 0.5, 0.5), "cm"),
                                gene.aes = aes(xmin = start, xmax = end, y = ymin, fill = strand, forward = orientation, label = gene),
                                #legend.position.prime='top',
                                #labels=labs(fill = "Strand"),
                                vline  = vline,  # if (exists('vline' )) vline  else NULL,
                                vline2 = vline2, # if (exists('vline2')) vline2 else NULL,
                                xlab_name='Genomic Position', ylab_name='5`-end coverage percent (%)',
                                return.plot.data=F, 
                                ...)
  
}

plotfun.genome.only   <- function(samples=NA, visfrom = NA, visto = NA, ylims.gene = c(-2, 4), y.multip = 3, breakseq = 1500,...) {
  
  FigY    <- plot.genome.region(visfrom=visfrom,	visto=visto, samples = samples,
                                facet_cropF = facet_nested(rows=vars(hpi), scales=scales, drop=T),
                                plot.data=plot.sum, #[plot.sum$strand == '+',], 
                                #[TIS.df.sum.all$tss == tss,],
                                add.genome.plot=T, genome.only = T,
                                facet_genes = NULL, facet_TR = NULL,
                                gene.plotdata=gene.plotdata, genome=genome, prime = prime,
                                sum.counts.in.window=T, bin_width = bin_width, add.all.pos=T, y.thresh=NA,
                                gene.label=T, gene_name_col = 'gene_name', 
                                add.unstranded=T, 
                                add.feature=T, 
                                force.all.gene.up.and.down = T,
                                force.all.gene.down=F, angle = angle,
                                add.cageTSS=F, cagefr.clust=cagefr.clust, cagefr.clust.dist=3,
                                crop.FALSE = crop.FALSE, scales = scales,
                                geom    = geom_area(aes(x=prime5, y = count, fill=strand, color = strand), position = 'identity', alpha=alpha$cov_geom, color = "black", linewidth=0.5),
                                #geom    = geom_point(aes(x=prime5, y = count, color=strand), alpha=alpha$cov_geom, size=0.1),
                                palette=palette, alpha=alpha, genomplot.scale=genomplot.scale, #genomplot.scale,
                                gene.sizes = gene.sizes, sizes = sizes,
                                y.multip   = y.multip, y.log10 = F, ylim=ylim, #ybreaks = -5:5, 
                                ylims.gene = ylims.gene,
                                force.gene.y = T, 
                                gene.label.col='black',
                                legend.position='top', plot.title = NA, #letters[2], #paste0(virus, '_', Ori),
                                breakseq=breakseq, 
                                margins=unit(c(0.5, 0.5, 0.5, 0.5), "cm"),
                                gene.aes = aes(xmin = start, xmax = end, y = ymin, fill = strand, forward = orientation, label = gene),
                                #legend.position.prime='top',
                                #labels=labs(fill = "Strand"),
                                vline  = if (exists('vline' )) vline  else NULL,
                                vline2 = if (exists('vline2')) vline2 else NULL,
                                xlab_name='Genomic Position', ylab_name='Read Count',
                                return.plot.data=F
                                , ...)
  
  
}
